File: go/cryptoutil/ecc.go
package cryptoutil

import (
	"crypto"
	"crypto/ecdh"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/asn1"
	"fmt"
	"github.com/jc-lab/jscp/go/payloadpb"
	"github.com/pkg/errors"
)

// OIDs for public key algorithms
var (
	oidPublicKeyRSA   = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 1, 1}
	oidPublicKeyECDSA = asn1.ObjectIdentifier{1, 2, 840, 10045, 2, 1}
)

// OIDs for elliptic curves
var (
	oidNamedCurveP224 = asn1.ObjectIdentifier{1, 3, 132, 0, 33}
	oidNamedCurveP256 = asn1.ObjectIdentifier{1, 2, 840, 10045, 3, 1, 7}
	oidNamedCurveP384 = asn1.ObjectIdentifier{1, 3, 132, 0, 34}
	oidNamedCurveP521 = asn1.ObjectIdentifier{1, 3, 132, 0, 35}
)

type curveDefine struct {
	Dsa elliptic.Curve
	Dh  ecdh.Curve
}

// Mapping of curve OIDs to curve implementations
var curveByOID = map[string]*curveDefine{
	oidNamedCurveP256.String(): &curveDefine{
		Dsa: elliptic.P256(),
		Dh:  ecdh.P256(),
	},

	oidNamedCurveP384.String(): &curveDefine{
		Dsa: elliptic.P384(),
		Dh:  ecdh.P384(),
	},

	oidNamedCurveP521.String(): &curveDefine{
		Dsa: elliptic.P521(),
		Dh:  ecdh.P521(),
	},
}

// SubjectPublicKeyInfo represents the ASN.1 structure of the same name
type SubjectPublicKeyInfo struct {
	Algorithm pkix.AlgorithmIdentifier
	PublicKey asn1.BitString
}

type ECCAlgorithm struct {
	keyUsage KeyUsage
}

type ECCDHSpecificAlgorithm struct {
	ECCAlgorithm
	dhCurve ecdh.Curve
}

func (a *ECCDHSpecificAlgorithm) GetType() payloadpb.DHAlgorithm {
	return payloadpb.DHAlgorithm_DHECC
}

func (a *ECCDHSpecificAlgorithm) Generate() (*DHKeyPair, error) {
	privateKey, err := a.dhCurve.GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	eccPrivateKey := NewECCPrivateKey(nil, privateKey, KeyUsageDH)

	return &DHKeyPair{
		Private: eccPrivateKey,
		Public:  eccPrivateKey.GetDHPublic(),
	}, nil
}

func (a *ECCDHSpecificAlgorithm) UnmarshalDHPublicKey(input []byte) (DHPublicKey, error) {
	return a.unmarshalPublicKey(input)
}

func (a *ECCAlgorithm) GetKeyFormat() payloadpb.KeyFormat {
	return payloadpb.KeyFormat_KeyFormatSubjectPublicKeyInfo
}

func (a *ECCAlgorithm) unmarshalPublicKey(input []byte) (*ECCPublicKey, error) {
	var spki SubjectPublicKeyInfo
	rest, err := asn1.Unmarshal(input, &spki)
	if err != nil {
		return nil, fmt.Errorf("failed to parse SubjectPublicKeyInfo: %w", err)
	}
	if len(rest) > 0 {
		return nil, errors.New("trailing data after SubjectPublicKeyInfo")
	}

	if !spki.Algorithm.Algorithm.Equal(oidPublicKeyECDSA) {
		return nil, errors.New("invalid ECC algorithm")
	}

	// ECDSA public key
	namedCurveOID := new(asn1.ObjectIdentifier)
	if _, err := asn1.Unmarshal(spki.Algorithm.Parameters.FullBytes, namedCurveOID); err != nil {
		return nil, fmt.Errorf("failed to parse EC parameters: %w", err)
	}

	namedCurve := curveByOID[namedCurveOID.String()]
	if namedCurve == nil {
		return nil, fmt.Errorf("unsupported elliptic curve: %v", namedCurveOID)
	}

	x, y := elliptic.Unmarshal(namedCurve.Dsa, spki.PublicKey.Bytes)
	if x == nil {
		return nil, errors.New("failed to unmarshal EC point")
	}

	dsaKey := &ecdsa.PublicKey{
		Curve: namedCurve.Dsa,
		X:     x,
		Y:     y,
	}
	dhKey, err := namedCurve.Dh.NewPublicKey(spki.PublicKey.Bytes)
	if err != nil {
		return nil, errors.Wrap(err, "failed to create dh key")
	}

	instance := &ECCPublicKey{
		keyUsage: a.keyUsage,
		DsaKey:   dsaKey,
		DhKey:    dhKey,
	}
	return instance, nil
}

func (a *ECCAlgorithm) UnmarshalPublicKey(input []byte) (PublicKey, error) {
	return a.unmarshalPublicKey(input)
}

type ECCPublicKey struct {
	DsaKey   *ecdsa.PublicKey
	DhKey    *ecdh.PublicKey
	keyUsage KeyUsage
}

func (k *ECCPublicKey) IsDHKey() bool {
	return k.keyUsage == KeyUsageDH
}

func (k *ECCPublicKey) IsSignatureKey() bool {
	return k.keyUsage == KeyUsageSignature
}

func (k *ECCPublicKey) Algorithm() PublicAlgorithm {
	return &x509Algorithm
}

func (k *ECCPublicKey) GetDHAlgorithm() DHAlgorithm {
	if k.DhKey == nil {
		return nil
	}
	return &ECCDHSpecificAlgorithm{
		dhCurve: k.DhKey.Curve(),
	}
}

func (k *ECCPublicKey) Verify(data []byte, signature []byte) (bool, error) {
	return ecdsa.VerifyASN1(k.DsaKey, Hash(data), signature), nil
}

func (k *ECCPublicKey) MarshalToProto() (*payloadpb.PublicKey, error) {
	var raw []byte
	var err error
	if k.keyUsage == KeyUsageSignature {
		raw, err = x509.MarshalPKIXPublicKey(k.DsaKey)
	} else if k.keyUsage == KeyUsageDH {
		raw, err = x509.MarshalPKIXPublicKey(k.DhKey)
	}
	if err != nil {
		return nil, err
	}
	return &payloadpb.PublicKey{
		Format: payloadpb.KeyFormat_KeyFormatSubjectPublicKeyInfo,
		Data:   raw,
	}, nil
}

func (k *ECCPublicKey) GetDHAlgorithmProto() payloadpb.DHAlgorithm {
	return payloadpb.DHAlgorithm_DHECC
}

func (k *ECCPublicKey) Marshal() []byte {
	p, _ := k.MarshalToProto()
	return p.Data
}

type ECCPrivateKey struct {
	DsaKey   *ecdsa.PrivateKey
	DhKey    *ecdh.PrivateKey
	keyUsage KeyUsage
}

func NewECCPrivateKey(ecdsaKey *ecdsa.PrivateKey, dhKey *ecdh.PrivateKey, keyUsage KeyUsage) *ECCPrivateKey {
	return &ECCPrivateKey{
		DsaKey:   ecdsaKey,
		DhKey:    dhKey,
		keyUsage: keyUsage,
	}
}

func (k *ECCPrivateKey) IsDHKey() bool {
	return k.keyUsage == KeyUsageDH
}

func (k *ECCPrivateKey) IsSignatureKey() bool {
	return k.keyUsage == KeyUsageSignature
}

func (k *ECCPrivateKey) Algorithm() PublicAlgorithm {
	return &x509Algorithm
}

func (k *ECCPrivateKey) GetDHAlgorithm() DHAlgorithm {
	if k.DhKey == nil {
		return nil
	}
	return &ECCDHSpecificAlgorithm{
		dhCurve: k.DhKey.Curve(),
	}
}

func (k *ECCPrivateKey) getPublic() *ECCPublicKey {
	p := &ECCPublicKey{
		keyUsage: k.keyUsage,
	}
	if k.DsaKey != nil {
		p.DsaKey = &k.DsaKey.PublicKey
	}
	if k.DhKey != nil {
		p.DhKey = k.DhKey.PublicKey()
	}
	return p
}

func (k *ECCPrivateKey) GetPublic() PublicKey {
	return k.getPublic()
}

func (k *ECCPrivateKey) GetSignaturePublicKey() SignaturePublicKey {
	return k.getPublic()
}

func (k *ECCPrivateKey) Sign(data []byte) ([]byte, error) {
	return k.DsaKey.Sign(rand.Reader, Hash(data), crypto.SHA256)
}

func (k *ECCPrivateKey) GetDHAlgorithmProto() payloadpb.DHAlgorithm {
	return payloadpb.DHAlgorithm_DHECC
}

func (k *ECCPrivateKey) GetDHPublic() DHPublicKey {
	return k.getPublic()
}

func (k *ECCPrivateKey) DH(peerKey DHPublicKey) ([]byte, error) {
	publicKey, ok := peerKey.(*ECCPublicKey)
	if !ok {
		return nil, errors.New("peerKey is not ECCDH")
	}
	return k.DhKey.ECDH(publicKey.DhKey)
}

File: node/src/crypto/ecc.ts
import * as pkijs from 'pkijs';
import * as asn1js from 'asn1js';
import { Buffer } from 'buffer';
import {
    PublicAlgorithm,
    DHAlgorithm,
    PublicKey,
    DHPublicKey,
    DHKeyPair,
    DHPrivateKey,
    SignaturePublicKey,
    SignaturePrivateKey,
} from './types';
import * as proto from '../proto';

import { type CurveFn } from '@noble/curves/abstract/weierstrass';

import { p256 } from '@noble/curves/p256';
import { p384 } from '@noble/curves/p384';
import { p521 } from '@noble/curves/p521';

export type CurveDefine = {
    oid: string;
} & CurveFn

export const CURVES: Record<string, CurveDefine> = {
    'P-256': {
        oid: '1.2.840.10045.3.1.7',
        ...p256,
    },
    'P-384': {
        oid: '1.3.132.0.34',
        ...p384,
    },
    'P-521': {
        oid: '1.3.132.0.35',
        ...p521,
    },
};

export enum KeyUsage {
    DH = 1,
    Signature = 2,
}

export interface SubjectPublicKeyInfo {
    algorithm: {
        algorithm: string;
        parameters: ArrayBuffer;
    };
    publicKey: ArrayBuffer;
}

export class ECCAlgorithm implements PublicAlgorithm {
    protected keyUsage: KeyUsage;

    constructor(keyUsage: KeyUsage) {
        this.keyUsage = keyUsage;
    }

    // PublicAlgorithm

    getKeyFormat(): proto.KeyFormat {
        return proto.KeyFormat.KeyFormatSubjectPublicKeyInfo;
    }

    async unmarshalPublicKey(input: Uint8Array): Promise<PublicKey> {
        return ECCAlgorithm.unmarshalPublicKeyImpl(this.keyUsage, input);
    }

    static async unmarshalPublicKeyImpl(keyUsage: KeyUsage, input: Uint8Array): Promise<ECCPublicKey> {
        const publicKeyInfo = pkijs.PublicKeyInfo.fromBER(input);
        if (publicKeyInfo.algorithm.algorithmId !== '1.2.840.10045.2.1') {
            throw new Error(`unknown public key: ${publicKeyInfo.algorithm.algorithmId}`);
        }
        const algorithmParams = publicKeyInfo.algorithm.algorithmParams;
        if (!(algorithmParams instanceof asn1js.ObjectIdentifier)) {
            throw new Error(`unknown public key: ${publicKeyInfo.algorithm}`);
        }
        const algorithmOid = algorithmParams.getValue();
        const curve = Object.values(CURVES).find(v => v.oid === algorithmOid);
        if (!curve) {
            throw new Error(`unknown curve: ${algorithmOid}`);
        }
        const keyRaw = publicKeyInfo.subjectPublicKey.valueBlock.valueHexView;
        return new ECCPublicKey(keyUsage, curve, keyRaw);
    }
}

export class ECCDHSpecificAlgorithm extends ECCAlgorithm implements DHAlgorithm {
    constructor(
        public readonly curve: CurveDefine
    ) {
        super(KeyUsage.DH);
    }

    // DHAlgorithm

    getType(): proto.DHAlgorithm {
        return proto.DHAlgorithm.DHECC;
    }

    async generate(): Promise<DHKeyPair> {
        const privateKey = this.curve.utils.randomPrivateKey();
        const publicKey = this.curve.getPublicKey(privateKey);
        return {
            private: new ECCPrivateKey(this.keyUsage, this.curve, privateKey),
            public: new ECCPublicKey(this.keyUsage, this.curve, publicKey),
        };
    }

    async unmarshalDHPublicKey(input: Uint8Array): Promise<DHPublicKey> {
        return ECCAlgorithm.unmarshalPublicKeyImpl(KeyUsage.DH, input)
    }
}

export class ECCPublicKey implements DHPublicKey, SignaturePublicKey {
    constructor(
        public readonly keyUsage: KeyUsage,
        public readonly curve: CurveDefine,
        public readonly key: Uint8Array,
    ) {
    }

    isDHKey(): boolean {
        return this.keyUsage === KeyUsage.DH;
    }

    isSignatureKey(): boolean {
        return this.keyUsage === KeyUsage.Signature;
    }

    algorithm(): PublicAlgorithm {
        return this.getAlgorithmImpl();
    }

    async verify(data: Uint8Array, signature: Uint8Array): Promise<boolean> {
        return this.curve.verify(signature, data, this.key, { prehash: true })
    }

    getDHAlgorithm(): DHAlgorithm {
        return this.getAlgorithmImpl();
    }

    getDHAlgorithmProto(): proto.DHAlgorithm {
        return proto.DHAlgorithm.DHECC;
    }

    marshal(): Uint8Array {
        const publicKeyInfo = new pkijs.PublicKeyInfo({
            algorithm: new pkijs.AlgorithmIdentifier({
                algorithmId: '1.2.840.10045.2.1',
                algorithmParams: new asn1js.ObjectIdentifier({
                    value: this.curve.oid,
                })
            }),
            subjectPublicKey: new asn1js.BitString({
                valueHex: this.key,
            })
        })
        return Buffer.from(publicKeyInfo.toString('base64'), 'base64');
    }

    marshalToProto(): proto.PublicKey {
        return {
            format: proto.KeyFormat.KeyFormatSubjectPublicKeyInfo,
            data: this.marshal(),
        };
    }

    getAlgorithmImpl(): ECCDHSpecificAlgorithm {
        return new ECCDHSpecificAlgorithm(this.curve);
    }
}

export class ECCPrivateKey implements DHPrivateKey, SignaturePrivateKey {
    constructor(
        public readonly keyUsage: KeyUsage,
        public readonly curve: CurveDefine,
        public readonly key: Uint8Array,
    ) {}

    isDHKey(): boolean {
        return this.keyUsage === KeyUsage.DH;
    }

    isSignatureKey(): boolean {
        return this.keyUsage === KeyUsage.Signature;
    }

    algorithm(): PublicAlgorithm {
        return this.getAlgorithmImpl();
    }

    toPublic(): ECCPublicKey {
        return new ECCPublicKey(this.keyUsage, this.curve, this.curve.getPublicKey(this.key, true));
    }

    getPublic(): PublicKey {
        return this.toPublic();
    }

    getDHPublic(): DHPublicKey {
        return this.toPublic();
    }

    getSignaturePublicKey(): SignaturePublicKey {
        return this.toPublic();
    }

    async sign(data: Uint8Array): Promise<Uint8Array> {
        return this.curve.sign(data, this.key, { prehash: true }).toDERRawBytes(true);
    }

    getDHAlgorithm(): DHAlgorithm {
        return this.getAlgorithmImpl();
    }

    getDHAlgorithmProto(): proto.DHAlgorithm {
        return proto.DHAlgorithm.DHECC;
    }

    async dh(peerKey: DHPublicKey): Promise<Uint8Array> {
        if (!(peerKey instanceof ECCPublicKey)) {
            throw new Error('peerKey is not ECCDH');
        }
        return this.curve.getSharedSecret(this.key, peerKey.key, true);
    }

    getAlgorithmImpl(): ECCDHSpecificAlgorithm {
        return new ECCDHSpecificAlgorithm(this.curve);
    }
}
